src/services/postgres_rfqt_firm_quote_validator.ts

```diff
@@ -0,0 +1,188 @@
+        // that can happen if two threads try to insert the same entry at the same time. This is why we add
+        // the "ON CONFLICT" clause.
+        if (makerAddressesToAddToCache.length > 0) {
+            logger.info(`Adding new addresses to cache: ${JSON.stringify(makerAddressesToAddToCache)}`);
+            await this._chainCacheRepository
+                .createQueryBuilder()
+                .insert()
+                .values(
+                    makerAddressesToAddToCache.map(makerAddress => {
+                        return {
+                            makerAddress,
+                            tokenAddress: makerTokenAddress,
+                            timeFirstSeen: 'NOW()',
+                        };
+                    }),
+                )
+                .onConflict(`("token_address", "maker_address") DO NOTHING`)
+                .execute();
+        }
+        return takerFillableAmounts;
+    }
+
+    private _calculateMakerBalanceFromResult(
+        result: MakerBalanceChainCacheEntity,
+        makerTokenAddress: string,
+        nowUnix: number,
+    ): BigNumber {
+        if (!result.timeOfSample) {
+            // If a record exists but a time of sample does not yet exist, this means that the cache entry has not yet been
+            // populated by the worker process. This may be due to a new address being added a few minutes ago, but it could
+            // also be due to a bug in the worker.
+            const timeFirstSeen = result.timeFirstSeen ? result.timeFirstSeen.getTime() : 0;
+            const msPassedSinceLastSeen = nowUnix - timeFirstSeen;
+            if (msPassedSinceLastSeen > this._cacheExpiryThresholdMs) {
+                logger.error(
+                    `Cache entry for maker ${result.makerAddress} and token ${result.tokenAddress} was first added on ${timeFirstSeen} which is more than ${this._cacheExpiryThresholdMs} ms ago. Assuming worker is stuck and setting maker balance to 0.`,
+                );
+                return ZERO;
+            } else {
+                logger.warn(
+                    `Cannot find cache for token ${makerTokenAddress} and maker ${result.makerAddress}. This entry was recently added so assuming the entire maker fillable amount is available`,
+                );
+                return new BigNumber(Number.POSITIVE_INFINITY);
+            }
+        } else if (nowUnix - result.timeOfSample.getTime() > this._cacheExpiryThresholdMs) {
+            // In this case a cache entry exists, but it's simply too old and this should never really happen unless the worker is stuck.
+            logger.error(
+                `Cache entry for maker ${result.makerAddress} and token ${
+                    result.tokenAddress
+                } was last refreshed on ${result.timeOfSample.getTime()} which is more than ${
+                    this._cacheExpiryThresholdMs
+                } ms ago. Assuming worker is stuck and setting maker balance to 0.`,
```

Comment:
This seems like a harsh assumption. Maybe we want to assume it's fillable in that case?

Reply:
discussed